leverage Fig's shell parser, add `git` spec

meganrogge opened the pull request:
fixes #239519
fixes #239518
Part of #239515

meganrogge committed:
add fig modules

meganrogge committed:
change tsconfig

meganrogge committed:
fix errors

meganrogge committed:
add log statement

meganrogge committed:
eslint, filters

meganrogge committed:
push current status

meganrogge committed:
fix things so there are no errors

Co-authored-by: Daniel Imms <Tyriar@users.noreply.github.com>

meganrogge committed:
get arg parsing to work

Co-authored-by: Daniel Imms <2193314+Tyriar@users.noreply.github.com>

Tyriar committed:
Merge remote-tracking branch 'origin/main' into merogge/fig-project

Tyriar committed:
Add tests for shell-parser

Tyriar committed:
Add readme explaining fork and cgmanifest

Tyriar committed:
Add upstream tests for shell-parser

Tyriar committed:
Remove additional src dir in forked fig packages

Tyriar committed:
Add fig/shared/ unit tests

Tyriar committed:
Remove unneeded files/content, move chmanifest to ext root

Tyriar committed:
Tweak folder names, rewrite require to import

Tyriar committed:
Pass in code spec

meganrogge committed:
get parsing to work for options/args

meganrogge committed:
Refactor

meganrogge committed:
add git for testing purposes

meganrogge committed:
get files and folders working

meganrogge committed:
clean up

meganrogge committed:
pass things into the tests

Tyriar committed:
Merge remote-tracking branch 'origin/main' into merogge/fig-project

Tyriar committed:
Use SuggestionFlag bitfield correctly

Tyriar committed:
Pull completions off Record

Tyriar committed:
Get basic generator working

meganrogge committed:
get tests passing

meganrogge committed:
clean up

meganrogge committed:
more cleanup

meganrogge committed:
rm comment

meganrogge committed:
rm comment

meganrogge committed:
push git tests, some failing

meganrogge committed:
comment out tests

meganrogge committed:
comment more out

meganrogge committed:
rm from filters

meganrogge committed:
use as upstream

meganrogge committed:
add helper

meganrogge committed:
better name

meganrogge committed:
clean up

meganrogge committed:
rm ?

meganrogge committed:
rm a bunch of code

meganrogge commented on extensions/terminal-suggest/src/terminalSuggestMain.ts:
```diff
@@ -316,15 +423,10 @@ export async function getCompletionItemsFromSpecs(
 		}
 		filesRequested = true;
 		foldersRequested = true;
```
I don't have the bandwidth today to think about this, but wonder if always setting these like this is now incorrect?

Tyriar replied:
https://github.com/microsoft/vscode/issues/240033

Tyriar committed:
Remove user aliases from test expectations

Tyriar committed:
Align tsconfig with typescript exts

Tyriar commented on extensions/terminal-suggest/src/completions/upstream/git.ts:
```diff
@@ -0,0 +1,9813 @@
+		script: [
+			"bash",
+			"-c",
+			"git --no-optional-locks status --short | sed -ne '/^M /p' -e '/A /p'",
```
There are 3 cases that use `| sed` which will fail currently on Windows. Created https://github.com/microsoft/vscode/issues/240031 to fix and upstream the change

Tyriar commented on extensions/terminal-suggest/src/fig/autocomplete-parser/constants.ts:
```diff
@@ -0,0 +1,33 @@
+const AWS_SPECS = ["aws", "q"];
+const UNIX_SPECS = [
+  "cd",
+  "git",
+  "rm",
+  "ls",
+  "cat",
+  "mv",
+  "ssh",
+  "cp",
+  "chmod",
+  "source",
+  "curl",
+  "make",
+  "mkdir",
+  "man",
+  "ln",
+  "grep",
+  "kill",
+];
+const EDITOR_SPECS = ["code", "nano", "vi", "vim", "nvim"];
+const JS_SPECS = ["node", "npm", "npx", "yarn"];
+const MACOS_SPECS = ["brew", "open"];
+const OTHER_SPECS = ["docker", "python"];
+
+export const MOST_USED_SPECS = [
+  ...AWS_SPECS,
+  ...UNIX_SPECS,
+  ...EDITOR_SPECS,
+  ...JS_SPECS,
+  ...MACOS_SPECS,
+  ...OTHER_SPECS,
+];
```
This might feed into their sorting?

Tyriar replied:
Just for loading: https://github.com/aws/amazon-q-developer-cli/blob/f66e0b0e917ab185eef528dc36eca56b78ca8b5d/packages/autocomplete-parser/src/loadHelpers.ts#L235-L246

Tyriar commented on extensions/terminal-suggest/src/test/completions/code.test.ts:
```diff
@@ -44,7 +44,7 @@ export function createCodeTestSpecs(executable: string): ITestSpec[] {
 		{ input: `${executable} --list-extensions |`, expectedCompletions: codeSpecOptions, expectedResourceRequests: { type: 'both', cwd: testPaths.cwd } },
 		{ input: `${executable} --show-versions |`, expectedCompletions: codeSpecOptions, expectedResourceRequests: { type: 'both', cwd: testPaths.cwd } },
 		{ input: `${executable} --category |`, expectedCompletions: categoryOptions },
-		{ input: `${executable} --category a|`, expectedCompletions: categoryOptions.filter(c => c.startsWith('a')) },
```
:+1:

Tyriar commented on extensions/terminal-suggest/src/test/completions/upstream/ls.test.ts:
```diff
@@ -66,12 +66,10 @@ export const lsTestSuiteSpec: ISuiteSpec = {
 		{ input: 'ls |', expectedCompletions: allOptions, expectedResourceRequests: { type: 'both', cwd: testPaths.cwd } },
 		{ input: 'ls -|', expectedCompletions: allOptions, expectedResourceRequests: { type: 'both', cwd: testPaths.cwd } },
 
-		// Filtering options should request all options so client side can filter
-		{ input: 'ls -a|', expectedCompletions: allOptions, expectedResourceRequests: { type: 'both', cwd: testPaths.cwd } },
```
If this is what fig gives us, we might need to make `-` a trigger character so that the list is refreshed after `-` and when one is deleted. https://github.com/microsoft/vscode/issues/240034

Tyriar commented on extensions/terminal-suggest/src/test/helpers.ts:
```diff
@@ -46,3 +46,15 @@ export function removeArrayEntries<T>(array: T[], ...elements: T[]): T[] {
 	return array;
 }
+
+export function getEnvAsRecord(shellIntegrationEnv: { [key: string]: string | undefined } | undefined): Record<string, string> {
+	const env: Record<string, string> = {};
+	if (shellIntegrationEnv) {
+		for (const [key, value] of Object.entries(shellIntegrationEnv)) {
+			if (typeof value === 'string') {
+				env[key] = value;
+			}
+		}
+	}
+	return env;
+}
```
We can maybe just do an `as` cast instead?

meganrogge committed:
Rm maybe unneeded file

Tyriar committed:
Merge branch 'main' into merogge/fig-project

Tyriar committed:
Remove test/ usage in src/

Tyriar committed:
Tweak eslint/filter exceptions

- third_party/ is ignored as it's not shipped
- fig/ is forked so we own the code and should convert to our style
- src/completions/upstream should be untouched upstream files

Tyriar committed:
Add ms copyright

Tyriar committed:
Use single quotes in src/fig/

Tyriar committed:
Normalize new lines when using test fixtures

Tyriar committed:
Comment out git test that can fail depending on .gitconfig

Tyriar committed:
Add comments to edit lines to index.d.ts

Tyriar commented on extensions/terminal-suggest/src/completions/index.d.ts:
```diff
@@ -1,1300 +1,1300 @@
+		 * @returns An array of `Suggestion` objects.
+		 *
+		 */
+		postProcess?: (out: string, tokens: string[]) => (Suggestion | null)[]; // <-- VS Code edit to make results correct
```
@meganrogge I added these comments to the 2 lines where we diverge to we can more easily track it

Tyriar committed:
Merge branch 'main' into merogge/fig-project

meganrogge committed:
add a line
